NetworkOnMainThread when reporting crash from notification (RB)

In Reference Browser whenever I click the "report" button in the crash notification then it immediately crashes again. I think I saw NetworkOnMainThread in the logs, but I couldn't really read them on my phone.

Just go this for Fenix too after reporting a crash from the notification ("report" action):

```
                         E  java.lang.RuntimeException: Unable to start service mozilla.components.lib.crash.service.SendCrashReportService@899fae9 with Intent { flg=0x10000000 cmp=org.mozilla.fennec_aurora/mozilla.components.lib.crash.service.SendCrashReportService bnds=[22,691][213,823] (has extras) }: android.os.N
                            etworkOnMainThreadException
                         E      at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4105)
                         E      at android.app.ActivityThread.access$1800(ActivityThread.java:219)
                         E      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1891)
                         E      at android.os.Handler.dispatchMessage(Handler.java:107)
                         E      at android.os.Looper.loop(Looper.java:214)
                         E      at android.app.ActivityThread.main(ActivityThread.java:7356)
                         E      at java.lang.reflect.Method.invoke(Native Method)
                         E      at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
                         E      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
                         E  Caused by: android.os.NetworkOnMainThreadException
                         E      at android.os.StrictMode$AndroidBlockGuardPolicy.onNetwork(StrictMode.java:1565)
                         E      at java.net.Inet6AddressImpl.lookupHostByName(Inet6AddressImpl.java:115)
                         E      at java.net.Inet6AddressImpl.lookupAllHostAddr(Inet6AddressImpl.java:103)
                         E      at java.net.InetAddress.getAllByName(InetAddress.java:1152)
                         E      at com.android.okhttp.Dns$1.lookup(Dns.java:41)
                         E      at com.android.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:178)
                         E      at com.android.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:144)
                         E      at com.android.okhttp.internal.http.RouteSelector.next(RouteSelector.java:86)
                         E      at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:176)
                         E      at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:128)
                         E      at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:97)
                         E      at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:289)
                         E      at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:232)
                         E      at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:465)
                         E      at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:131)
                         E      at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:262)
                         E      at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getOutputStream(DelegatingHttpsURLConnection.java:219)
                         E      at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:30)
                         E      at mozilla.components.lib.crash.service.MozillaSocorroService.sendReport$lib_crash_release(MozillaSocorroService.kt:10)
                         E      at mozilla.components.lib.crash.service.MozillaSocorroService.report(MozillaSocorroService.kt:2)
                         E      at mozilla.components.lib.crash.CrashReporter.submitReport(CrashReporter.kt:4)
                         E      at mozilla.components.lib.crash.service.SendCrashReportService.onStartCommand(SendCrashReportService.kt:23)
                         E      at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4087)
                         E      ... 8 more
```

--- a/components/lib/crash/src/main/java/mozilla/components/lib/crash/CrashReporter.kt
+++ b/components/lib/crash/src/main/java/mozilla/components/lib/crash/CrashReporter.kt
@@ -9,10 +9,12 @@ import android.content.Context
 import android.content.Intent
 import androidx.annotation.StyleRes
 import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.Job
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
 import mozilla.components.lib.crash.handler.ExceptionHandler
 import mozilla.components.lib.crash.notification.CrashNotification
 import mozilla.components.lib.crash.prompt.CrashPrompt
@@ -50,7 +52,8 @@ class CrashReporter(
     private val shouldPrompt: Prompt = Prompt.NEVER,
     var enabled: Boolean = true,
     internal val promptConfiguration: PromptConfiguration = PromptConfiguration(),
-    private val nonFatalCrashIntent: PendingIntent? = null
+    private val nonFatalCrashIntent: PendingIntent? = null,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
 ) {
     internal val logger = Logger("mozac/CrashReporter")
     internal val crashBreadcrumbs = BreadcrumbPriorityQueue(BREADCRUMB_MAX_NUM)
@@ -76,18 +79,21 @@ class CrashReporter(
 
     /**
      * Submit a crash report to all registered services.
-     *
-     * Note: This method may block and perform I/O on the calling thread.
      */
-    fun submitReport(crash: Crash) {
-        services.forEach { service ->
-            when (crash) {
-                is Crash.NativeCodeCrash -> service.report(crash)
-                is Crash.UncaughtExceptionCrash -> service.report(crash)
+    fun submitReport(crash: Crash, then: () -> Unit = {}): Job {
+        return scope.launch {
+            services.forEach { service ->
+                when (crash) {
+                    is Crash.NativeCodeCrash -> service.report(crash)
+                    is Crash.UncaughtExceptionCrash -> service.report(crash)
+                }
+            }
+
+            logger.info("Crash report submitted to ${services.size} services")
+            withContext(Dispatchers.Main) {
+                then()
             }
         }
-
-        logger.info("Crash report submitted to ${services.size} services")
     }
 
     /**
@@ -133,7 +139,6 @@ class CrashReporter(
             }
         } else {
             logger.info("Immediately submitting crash report")
-
             submitReport(crash)
         }
     }

--- a/components/lib/crash/src/main/java/mozilla/components/lib/crash/prompt/CrashReporterActivity.kt
+++ b/components/lib/crash/src/main/java/mozilla/components/lib/crash/prompt/CrashReporterActivity.kt
@@ -16,10 +16,6 @@ import kotlinx.android.synthetic.main.mozac_lib_crash_crashreporter.messageView
 import kotlinx.android.synthetic.main.mozac_lib_crash_crashreporter.restartButton
 import kotlinx.android.synthetic.main.mozac_lib_crash_crashreporter.sendCheckbox
 import kotlinx.android.synthetic.main.mozac_lib_crash_crashreporter.titleView
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.GlobalScope
-import kotlinx.coroutines.launch
-import kotlinx.coroutines.withContext
 import mozilla.components.lib.crash.Crash
 import mozilla.components.lib.crash.CrashReporter
 import mozilla.components.lib.crash.R
@@ -101,12 +97,8 @@ class CrashReporterActivity : AppCompatActivity() {
             return
         }
 
-        GlobalScope.launch(reporterCoroutineContext) {
-            crashReporter.submitReport(crash)
-
-            withContext(Dispatchers.Main) {
-                then()
-            }
+        crashReporter.submitReport(crash) {
+            then()
         }
     }
 

--- a/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/SendCrashReportService.kt
+++ b/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/SendCrashReportService.kt
@@ -9,13 +9,8 @@ import android.content.Context
 import android.content.Intent
 import android.os.Build
 import android.os.IBinder
-import androidx.annotation.VisibleForTesting
 import androidx.core.app.NotificationCompat
 import androidx.core.app.NotificationManagerCompat
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.GlobalScope
-import kotlinx.coroutines.launch
-import kotlinx.coroutines.withContext
 import mozilla.components.lib.crash.Crash
 import mozilla.components.lib.crash.CrashReporter
 import mozilla.components.lib.crash.R
@@ -23,12 +18,9 @@ import mozilla.components.lib.crash.notification.CrashNotification
 import mozilla.components.lib.crash.notification.NOTIFICATION_ID
 import mozilla.components.lib.crash.notification.NOTIFICATION_TAG
 import mozilla.components.support.base.ids.SharedIdsHelper
-import kotlin.coroutines.CoroutineContext
-import kotlin.coroutines.EmptyCoroutineContext
 
 class SendCrashReportService : Service() {
     private val crashReporter: CrashReporter by lazy { CrashReporter.requireInstance }
-    private var reporterCoroutineContext: CoroutineContext = EmptyCoroutineContext
 
     override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
@@ -48,21 +40,11 @@ class SendCrashReportService : Service() {
         }
 
         NotificationManagerCompat.from(this).cancel(NOTIFICATION_TAG, NOTIFICATION_ID)
-        crashReporter.submitReport(Crash.fromIntent(intent))
-        stopSelf()
+        crashReporter.submitReport(Crash.fromIntent(intent)) {
+            stopSelf()
+        }
 
         return START_NOT_STICKY
-    }
-
-    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-    internal fun sendCrashReport(crash: Crash, then: () -> Unit) {
-        GlobalScope.launch(reporterCoroutineContext) {
-            crashReporter.submitReport(crash)
-
-            withContext(Dispatchers.Main) {
-                then()
-            }
-        }
     }
 
     override fun onBind(intent: Intent): IBinder? {

--- a/components/lib/crash/src/test/java/mozilla/components/lib/crash/CrashReporterTest.kt
+++ b/components/lib/crash/src/test/java/mozilla/components/lib/crash/CrashReporterTest.kt
@@ -8,6 +8,8 @@ import android.app.Activity
 import android.app.PendingIntent
 import android.content.Intent
 import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
 import mozilla.components.lib.crash.service.CrashReporterService
 import mozilla.components.support.test.any
 import mozilla.components.support.test.eq
@@ -15,14 +17,16 @@ import mozilla.components.support.test.expectException
 import mozilla.components.support.test.ext.joinBlocking
 import mozilla.components.support.test.mock
 import mozilla.components.support.test.robolectric.testContext
+import mozilla.components.support.test.rule.MainCoroutineRule
 import org.junit.Assert.assertEquals
-import org.junit.Assert.assertFalse
 import org.junit.Assert.assertNotEquals
 import org.junit.Assert.assertNotNull
 import org.junit.Assert.assertTrue
 import org.junit.Before
+import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
+import org.mockito.Mockito.doReturn
 import org.mockito.Mockito.never
 import org.mockito.Mockito.spy
 import org.mockito.Mockito.verify
@@ -32,6 +36,10 @@ import java.lang.reflect.Modifier
 
 @RunWith(AndroidJUnit4::class)
 class CrashReporterTest {
+
+    @ExperimentalCoroutinesApi
+    @get:Rule
+    val coroutinesTestRule = MainCoroutineRule()
 
     @Before
     fun setUp() {
@@ -165,28 +173,32 @@ class CrashReporterTest {
     }
 
     @Test
-    fun `CrashReporter forwards crashes to service`() {
-        var nativeCrash = false
+    fun `CrashReporter submits crashes`() {
+        val crash = mock<Crash>()
+        val service = mock<CrashReporterService>()
+        val reporter = spy(CrashReporter(
+            services = listOf(service),
+            shouldPrompt = CrashReporter.Prompt.NEVER
+        ).install(testContext))
+
+        doReturn(mock<Job>()).`when`(reporter).submitReport(crash)
+        reporter.onCrash(testContext, crash)
+        verify(reporter).submitReport(crash)
+    }
+
+    @Test
+    fun `CrashReporter forwards uncaught exception crashes to service`() {
         var exceptionCrash = false
-        var caughtException = false
 
         val service = object : CrashReporterService {
             override fun report(crash: Crash.UncaughtExceptionCrash) {
                 exceptionCrash = true
-                nativeCrash = false
-                caughtException = false
             }
 
             override fun report(crash: Crash.NativeCodeCrash) {
-                exceptionCrash = false
-                nativeCrash = true
-                caughtException = false
             }
 
             override fun report(throwable: Throwable) {
-                exceptionCrash = false
-                nativeCrash = false
-                caughtException = true
             }
         }
 
@@ -195,28 +207,65 @@ class CrashReporterTest {
             shouldPrompt = CrashReporter.Prompt.NEVER
         ).install(testContext))
 
-        reporter.onCrash(
-            mock(),
-            Crash.UncaughtExceptionCrash(RuntimeException(), arrayListOf()))
+        reporter.submitReport(
+            Crash.UncaughtExceptionCrash(RuntimeException(), arrayListOf())
+        ).joinBlocking()
+        assertTrue(exceptionCrash)
+    }
+
+    @Test
+    fun `CrashReporter forwards native crashes to service`() {
+        var nativeCrash = false
+
+        val service = object : CrashReporterService {
+            override fun report(crash: Crash.UncaughtExceptionCrash) {
+            }
+
+            override fun report(crash: Crash.NativeCodeCrash) {
+                nativeCrash = true
+            }
+
+            override fun report(throwable: Throwable) {
+            }
+        }
+
+        val reporter = spy(CrashReporter(
+            services = listOf(service),
+            shouldPrompt = CrashReporter.Prompt.NEVER
+        ).install(testContext))
+
+        reporter.submitReport(
+            Crash.NativeCodeCrash("", true, "", false, arrayListOf())
+        ).joinBlocking()
+        assertTrue(nativeCrash)
+    }
+
+    @Test
+    fun `CrashReporter forwards caught exception crashes to service`() {
+        var exceptionCrash = false
+
+        val service = object : CrashReporterService {
+            override fun report(crash: Crash.UncaughtExceptionCrash) {
+            }
+
+            override fun report(crash: Crash.NativeCodeCrash) {
+            }
+
+            override fun report(throwable: Throwable) {
+                exceptionCrash = true
+            }
+        }
+
+        val reporter = spy(CrashReporter(
+            services = listOf(service),
+            shouldPrompt = CrashReporter.Prompt.NEVER
+        ).install(testContext))
+
+        reporter.submitCaughtException(
+            RuntimeException()
+        ).joinBlocking()
 
         assertTrue(exceptionCrash)
-        assertFalse(nativeCrash)
-        assertFalse(caughtException)
-
-        reporter.onCrash(
-            mock(),
-            Crash.NativeCodeCrash("", true, "", false, arrayListOf())
-        )
-
-        assertFalse(exceptionCrash)
-        assertTrue(nativeCrash)
-        assertFalse(caughtException)
-
-        reporter.submitCaughtException(RuntimeException()).joinBlocking()
-
-        assertFalse(exceptionCrash)
-        assertFalse(nativeCrash)
-        assertTrue(caughtException)
     }
 
     @Test

--- a/components/lib/crash/src/test/java/mozilla/components/lib/crash/handler/CrashHandlerServiceTest.kt
+++ b/components/lib/crash/src/test/java/mozilla/components/lib/crash/handler/CrashHandlerServiceTest.kt
@@ -7,21 +7,32 @@ package mozilla.components.lib.crash.handler
 import android.content.ComponentName
 import android.content.Intent
 import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.TestCoroutineScope
 import mozilla.components.lib.crash.Crash
 import mozilla.components.lib.crash.CrashReporter
 import mozilla.components.lib.crash.service.CrashReporterService
 import mozilla.components.support.test.robolectric.testContext
+import mozilla.components.support.test.rule.MainCoroutineRule
 import org.junit.After
 import org.junit.Assert.assertEquals
 import org.junit.Assert.assertNotNull
 import org.junit.Assert.fail
+import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.Mockito.doNothing
 import org.mockito.Mockito.spy
 
+@ExperimentalCoroutinesApi
 @RunWith(AndroidJUnit4::class)
 class CrashHandlerServiceTest {
+
+    private val testDispatcher = TestCoroutineDispatcher()
+
+    @get:Rule
+    val coroutinesTestRule = MainCoroutineRule(testDispatcher)
 
     @After
     fun tearDown() {
@@ -32,6 +43,7 @@ class CrashHandlerServiceTest {
     fun `CrashHandlerService will forward GeckoView crash to crash reporter`() {
         var caughtCrash: Crash.NativeCodeCrash? = null
 
+        val scope = TestCoroutineScope(testDispatcher)
         CrashReporter(
             shouldPrompt = CrashReporter.Prompt.NEVER,
             services = listOf(object : CrashReporterService {
@@ -46,7 +58,8 @@ class CrashHandlerServiceTest {
                 override fun report(throwable: Throwable) {
                     fail("Didn't expect caught exception")
                 }
-            })
+            }),
+            scope = scope
         ).install(testContext)
 
         val intent = Intent("org.mozilla.gecko.ACTION_CRASHED")
@@ -67,7 +80,9 @@ class CrashHandlerServiceTest {
 
         val service = spy(CrashHandlerService())
         doNothing().`when`(service).kill()
+
         service.onHandleIntent(intent)
+        scope.advanceUntilIdle()
 
         assertNotNull(caughtCrash)
 

--- a/components/lib/crash/src/test/java/mozilla/components/lib/crash/service/SendCrashReportServiceTest.kt
+++ b/components/lib/crash/src/test/java/mozilla/components/lib/crash/service/SendCrashReportServiceTest.kt
@@ -9,6 +9,7 @@ import android.content.Intent
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import mozilla.components.lib.crash.Crash
 import mozilla.components.lib.crash.CrashReporter
+import mozilla.components.support.test.any
 import mozilla.components.support.test.eq
 import mozilla.components.support.test.robolectric.testContext
 import org.junit.After
@@ -77,6 +78,6 @@ class SendCrashReportServiceTest {
         originalCrash.fillIn(intent)
 
         service?.onStartCommand(intent, 0, 0)
-        verify(crashReporter).submitReport(eq(originalCrash))
+        verify(crashReporter).submitReport(eq(originalCrash), any())
     }
 }

--- a/docs/changelog.md
+++ b/docs/changelog.md
@@ -50,6 +50,14 @@ permalink: /changelog/
 * **browser-menu**
   * Added `BrowserMenuHighlightableSwitch` to represent a highlightable item with a toggle switch.
 
+* **lib-crash**
+  * Now supports performing action after submitting crash report.
+  ```kotlin
+  crashReporter.submitReport(Crash.fromIntent(intent)) {
+      stopSelf()
+  }
+  ```
+
 * **support-ktx**
   * Added `Context.getDrawableWithTint` extension method to get a drawable resource with a tint applied.
 	* `String.isUrl` is now using a more lenient check for improved performance. Strictly checking whether a string is a URL or not is supported through the new `String.isUrlStrict` method.

